From 5d0a8ab50dc99ed880c7d56be3a206557e03536d Mon Sep 17 00:00:00 2001
From: Mao Jinlong <jinlmao@codeaurora.org>
Date: Tue, 20 Nov 2018 16:11:13 +0800
Subject: [PATCH] DeskClock: Use POWER_OFF_ALARM permission

Use POWER_OFF_ALARM permission to send set and cancel power off alarm
actions.

CRs-fixed: 2282053
Change-Id: I1a7984cb90db289d1bd7377963dbd28aa866eb10
---
 AndroidManifest.xml                      |  1 +
 src/com/android/deskclock/DeskClock.java | 24 ++++++++++++++++++++++++
 2 files changed, 25 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index fdc9ef245..9cfa6ef77 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -29,6 +29,7 @@
     <uses-permission android:name="android.permission.VIBRATE" />
     <uses-permission android:name="android.permission.DISABLE_KEYGUARD" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <uses-permission android:name="org.codeaurora.permission.POWER_OFF_ALARM" />
 
     <!-- WRITE_SETTINGS is required to record the upcoming alarm prior to L -->
     <uses-permission
diff --git a/src/com/android/deskclock/DeskClock.java b/src/com/android/deskclock/DeskClock.java
index 5585956ee..afee29d54 100644
--- a/src/com/android/deskclock/DeskClock.java
+++ b/src/com/android/deskclock/DeskClock.java
@@ -22,6 +22,7 @@ import android.animation.AnimatorSet;
 import android.animation.ValueAnimator;
 import android.app.Fragment;
 import android.content.Intent;
+import android.content.pm.PackageManager;
 import android.graphics.drawable.Drawable;
 import android.os.Bundle;
 import android.support.annotation.StringRes;
@@ -49,6 +50,7 @@ import com.android.deskclock.data.DataModel;
 import com.android.deskclock.data.DataModel.SilentSetting;
 import com.android.deskclock.data.OnSilentSettingsListener;
 import com.android.deskclock.events.Events;
+import com.android.deskclock.LogUtils;
 import com.android.deskclock.provider.Alarm;
 import com.android.deskclock.uidata.TabListener;
 import com.android.deskclock.uidata.UiDataModel;
@@ -128,6 +130,11 @@ public class DeskClock extends BaseActivity
     /** {@code true} when a settings change necessitates recreating this activity. */
     private boolean mRecreateActivity;
 
+    private static final String PERMISSION_POWER_OFF_ALARM =
+            "org.codeaurora.permission.POWER_OFF_ALARM";
+
+    private static final int CODE_FOR_ALARM_PERMISSION = 1;
+
     @Override
     public void onNewIntent(Intent newIntent) {
         super.onNewIntent(newIntent);
@@ -143,6 +150,8 @@ public class DeskClock extends BaseActivity
         setContentView(R.layout.desk_clock);
         mSnackbarAnchor = findViewById(R.id.content);
 
+        checkPermissions();
+
         // Configure the toolbar.
         final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
         setSupportActionBar(toolbar);
@@ -463,6 +472,21 @@ public class DeskClock extends BaseActivity
         }
     }
 
+    private void checkPermissions() {
+        if (checkSelfPermission(PERMISSION_POWER_OFF_ALARM)
+                != PackageManager.PERMISSION_GRANTED) {
+            requestPermissions(new String[]{PERMISSION_POWER_OFF_ALARM}, CODE_FOR_ALARM_PERMISSION);
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           String permissions[], int[] grantResults) {
+        if (requestCode == CODE_FOR_ALARM_PERMISSION){
+            LogUtils.i("Power off alarm permission is granted.");
+        }
+    }
+
     /**
      * Configure the {@link #mFragmentTabPager} and {@link #mTabLayout} to display UiDataModel's
      * selected tab.
-- 
2.17.1

